workshop-1.7.py

```diff
@@ -4,65 +4,60 @@
 #
 # Run: bokeh serve workshop-1.7.py
 
-import json
-import operator
 import colormap
-from parking.meter import Meter
-from parking.transaction import Transaction
 from itertools import groupby
-from functools import reduce
-from math import floor, ceil
-from gmapplot import GoogleMapPlot
-
-
-SAN_DIEGO_COORDINATE = (32.712, -117.1611)
-DEFAULT_ZOOM = 16
-
-HEATMAP_GRANULARITY = 10
-# Scale the data upwards to get better coloration
-HEATMAP_SCALING = 3
+from math import floor
 
-
-# Read the config.json file
-config = []
-with open('config.json', 'r') as f:
-	config = json.load(f)
+from config import get_config
+from constants import DEFAULT_ZOOM, HEATMAP_GRANULARITY, HEATMAP_SCALING, SAN_DIEGO_COORDINATE
+from gmapplot import GoogleMapPlot
+from parking.meter import Meter
+from parking.transaction import Transaction
 
 
 # Google requires you obtain and enable an API key:
 #
 #     https://developers.google.com/maps/documentation/javascript/get-api-key
 #
 # Replace the value in config.json below with your personal API key:
-assert 'GoogleMapsAPIKey' in config and len(config['GoogleMapsAPIKey']) > 0, "Add your own Google Maps API Key to config.json"
-plot = GoogleMapPlot(config['GoogleMapsAPIKey'], SAN_DIEGO_COORDINATE[0], SAN_DIEGO_COORDINATE[1], "roadmap", DEFAULT_ZOOM)
-        
+config = get_config()
+plot = GoogleMapPlot(
+    api_key=config['GoogleMapsAPIKey'],
+    lat=SAN_DIEGO_COORDINATE[0],
+    lng=SAN_DIEGO_COORDINATE[1],
+    type="roadmap",
+    zoom=DEFAULT_ZOOM
+)
+
 
 # Load the parking meter transactions before we start
 Transaction.transactions()
 poles = Meter.poles_in_gaslamp()
 color_map = colormap.color_map(HEATMAP_GRANULARITY)
 
-datasources = [plot.draw_points_with_circle_glyph([], { 'fill_color' : color_map[i],
-							  							'size' : 4 }) for i in range(HEATMAP_GRANULARITY)]
+datasources = [plot.draw_points_with_circle_glyph([], attrs={'fill_color': color_map[i], 'size': 4})
+               for i in range(HEATMAP_GRANULARITY)]
 
 
 def update_plot(day):
-	transactions = Transaction.transactions_for_day(day)
+    transactions = Transaction.transactions_for_day(day)
 
-	poles_by_heat = [[] for i in range(HEATMAP_GRANULARITY)]
-	for k, g in groupby(transactions, lambda t: t.pole):
-		hours = reduce(operator.add, map(lambda t: t.duration, g), 0)
-		i = min(floor(HEATMAP_SCALING * hours / Transaction.MAX_DURATION * HEATMAP_GRANULARITY), HEATMAP_GRANULARITY - 1)
-		poles_by_heat[i] += [k]
+    poles_by_heat = [[] for i in range(HEATMAP_GRANULARITY)]
+    for pole, grouped_transactions in groupby(transactions, lambda t: t.pole):
+        hours = sum([t.duration for t in grouped_transactions])
+        i = min(
+            int(floor(HEATMAP_SCALING * hours / Transaction.MAX_DURATION * HEATMAP_GRANULARITY)),
```

Comment:
I accidentally got this going in Python 2...this (`floor` returning a float instead of an int) was the only part that broke on ye olde Python, so I just added this for full support.

Reply:
Cool. We're headed towards Python 3 (and I called that out in the workshop) because Jupyter Notebooks recommends it.

Reply:
Yeah everyone should be using Python 3 now. I'll back this out.